### PR TITLE
Self funded wallet derivation

### LINKED ISSUE #179: Support self-funded sponsor derivation
See: https://api3workspace.slack.com/archives/C05TQPT7PNJ/p1706521432528779.

Agreed on call that this is the best way to support self-funded dAPIs. Use the derivation that Nodary uses and add a setting in the config which chooses the right option. Don't forget to document it in the README.

### PR DESCRIPTION
**Pull request by Siegrift**
Closes https://github.com/api3dao/airseeker-v2/issues/179

## Rationale

We have not discussed the details (mostly naming) so I went ahead with the following:
1. Added `walletDerivationScheme` in the root of the `airseeker.json` config.
2. The supported values are `self-funded` and `managed`.
See the [README.md changes](https://github.com/api3dao/airseeker-v2/compare/read-feeds-values-from-chain...self-funded-wallet-derivation?expand=1#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5) for details.

The wallet derivation changes are [here](https://github.com/api3dao/airseeker-v2/compare/read-feeds-values-from-chain...self-funded-wallet-derivation?expand=1#diff-39b2554fd18da165b59a6351b1aafff3714e2a80c1435f2de9706355b4d32351) and the self-funded sponsor derivation takes into account the update parameters as referenced [here](https://api3workspace.slack.com/archives/C05TQPT7PNJ/p1706601083254309?thread_ts=1706521432.528779&cid=C05TQPT7PNJ).

## TODO

- [x] Mention this change in the spec.

### CODE REVIEW

File: local-test-configuration/monitoring/index.html [hunk 1]
```diff
@@ -858,6 +865,23 @@ <h2>Active data feeds</h2>
       return sponsorWallet;
     };
 
+    const deriveSponsorWallet = (
+      sponsorWalletMnemonic,
+      dapiNameOrDataFeedId,
+      updateParameters,
+      walletDerivationScheme
+    ) => {
+      // Derive the hash of the data feed using which the sponsor wallet is derived. For self-funded feeds it's more
```
Comment (aquarat) [inline local-test-configuration/monitoring/index.html]: This is a bit strange, but it's super minor (especially seeing as it's part of the _monitoring_ for the test configuration).

File: src/utils.ts [hunk 2]
```diff
@@ -43,6 +53,23 @@ export const deriveSponsorWallet = (sponsorWalletMnemonic: string, dapiNameOrDat
   return sponsorWallet;
 };
 
+export const deriveSponsorWallet = (
+  sponsorWalletMnemonic: string,
+  dapiNameOrDataFeedId: string,
+  updateParameters: string,
+  walletDerivationScheme: WalletDerivationScheme
+) => {
+  // Derive the hash of the data feed using which the sponsor wallet is derived. For self-funded feeds it's more
```
Comment (aquarat) [inline src/utils.ts]: ah and here, I'm guessing this was probably copied from somewhere.

**Review by aquarat — APPROVED** [review]
LGTM (again 🤣 ) 🚀

File: local-test-configuration/monitoring/index.html [hunk 3]
```diff
@@ -858,6 +865,23 @@ <h2>Active data feeds</h2>
...
+      // suitable to derive the hash also from update parameters. This does not apply to mananaged feeds which want to be
+      // funded by the same wallet independently of the update parameters.
+      const sponsorAddressHash =
```
Comment (bbenligiray) [inline local-test-configuration/monitoring/index.html:879]: This is the sponsor address, not the sponsor address hash, no?

File: README.md [hunk 4]
```diff
@@ -224,6 +224,15 @@ The fetch interval in seconds between retrievals of signed API data.
 
 A list of signed API URLs to call along with URLs fetched from the chain.
 
+#### `walletDerivationScheme`
+
+The derivation scheme used to derive sponsor wallets. The following options are available:
```
Comment (bbenligiray) [inline README.md]: Doesn't feel future compatible with https://api3workspace.slack.com/archives/C05TQPT7PNJ/p1706712947591099?thread_ts=1706521432.528779&cid=C05TQPT7PNJ

Comment (Siegrift) [reply README.md]: Yeah, but does it really need to be? I don't think it's a big issue if the Airseeker config changes and I wanted to prioritise clarity for current implementation. That said, if you want it future proof, we might do and object right away:
- `{ type: "managed" }`
- `{ type: "self-funded" }`
- (In the future) `{ type: "fixed-wallet", "mnemonic": "..." }` (mnemonic may not be needed if we use the Airseeker one in which case we could get away with just union of strings).

File: local-test-configuration/monitoring/index.html (hunk 3 above)
Comment (Siegrift) [reply local-test-configuration/monitoring/index.html:879]: No, I've separated the hashing part into separate functions and one base function (`deriveSponsorWalletFromSponsorAddressHash`) which takes first 20 bytes from the hash and interprets that as sponsor address and derives the wallet.

File: src/utils.ts (hunk 2 above)
Comment (Siegrift) [reply src/utils.ts]: Yeah. Basically, all the functions in the local monitoring are copied from the implementation (and maybe slightly updated) so there is just a single HTML file that can be opened in browser.

File: local-test-configuration/monitoring/index.html (hunk 1 above)
Comment (Siegrift) [reply local-test-configuration/monitoring/index.html]: Yeah, the comment is a bit confusing. I've tried rewording it to better reflect what's implemented. Let me know if it makes sense.

File: local-test-configuration/monitoring/index.html (hunk 3 above)
Comment (bbenligiray) [reply local-test-configuration/monitoring/index.html:879]: I mean the thing that you split into chunks and build the path out of is called the sponsor address, regardless if it's the address of an actual sponsor (such as [these](https://github.com/api3dao/operations-database/blob/48676e0e79114b5250f21cf7518100b0eb5e6645/src/scripts/contants.ts))

File: README.md (hunk 4 above)
Comment (bbenligiray) [reply README.md]: Yeah, add the `{}`, it's definitely worth it

File: local-test-configuration/monitoring/index.html (hunk 3 above)
Comment (Siegrift) [reply local-test-configuration/monitoring/index.html:879]: Yeah, that's why I don't want to name it `sponsorAddress`. There are:
- `sponsorAddressHash` - Hash from which we get the build path - e.g. `ethers.keccak256(ethers.solidityPacked(['bytes32', 'bytes'], [dapiNameOrDataFeedId, updateParameters]))`
- `sponsorAddress` - `ethers.getAddress(sponsorAddressHash.slice(0, 42))`
- `sponsorWallet` - Derived from mnemonic and `sponsorAddress`

Not sure if we are not misunderstanding each other. LMK if I am missing something.

File: README.md (hunk 4 above)
Comment (Siegrift) [reply README.md]: Done.

File: local-test-configuration/monitoring/index.html (hunk 3 above)
Comment (bbenligiray) [reply local-test-configuration/monitoring/index.html:879]: Yeah okay, I hadn't dug deep enough to find `sponsorAddress`

**Pull request merged**